protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+  - Example 1: The monorepo is at `op-contracts/v1.5.0`. Various tech debt is cleaned up in contracts, but no features are added, so all contracts only bump the patch version. The next monorepo release will be `op-contracts/v1.5.1`.
+  - Example 2: The monorepo is at `op-contracts/v1.5.1`. Legacy `ALL_CAPS()` getter methods are removed from a contract, causing that contract to bump the major version. The next monorepo release will be `op-contracts/v2.0.0`.
+- A monorepo contracts release must map to an exact set of contract semvers, and this mapping must be defined in the contract release notes which are the source of truth. See [`op-contracts/v1.4.0-rc.4`](https://github.com/ethereum-optimism/optimism/releases/tag/op-contracts%2Fv1.4.0-rc.4) for an example of what release notes should look like.
+
+The above gets us most of the way there, and there is one problem not yet addressed: When a release is proposed to governance, we propose a commit hash, and often those proposed contracts are already deployed to mainnet.
+For example, the [Fault Proofs governance proposal](https://gov.optimism.io/t/upgrade-proposal-fault-proofs/8161) provides specific addresses that will be used.
+This is ideal and we should continue to do that, but we need to fit this in to the above scheme.
+
+To accommodate this, once ready for governance approval the release flow is:
+
+- On the `develop` branch, bump the version of all contracts to their respective `X.Y.Z-rc.n`. The `X.Y.Z` here refers to the contract-specific versions so differs per-contract, whereas the `.rc.n` is the same across all contracts.
+- Branch off of `develop` and create a branch name `proposal/op-contracts/vX.Y.Z`. Here, `X.Y.Z` is the new version of the monorepo release.
+- On this branch, create a new commit that removes the `-rc.n` from all contracts. This is the commit hash that will be used to deploy the contracts and proposed to governance.
+- Once governance approves the proposal, merge the proposal branch into `develop` and bump the version of all contracts to their respective `X.Y.Z`. When there are conflicts, resolve them in favor of the `develop` branch. In other words:
+  - If a contract has not had any changes since the proposal, the version should be bumped to `X.Y.Z`.
+  - If a contract has had changes since the proposal, the version should be stay as whatever is the latest version on `develop`.
+
+Lastly, we will start maintaining a CHANGELOG for contract releases.
+The next expected release version will have a tracking issue that documents the new versions of each contract that will be included in the release, along with the links to the PRs that made the changes.
+Every contracts PR should have an accompanying changelog entry in the tracking issue once it is merged.
+See [ethereum-optimism/optimism#10592](https://github.com/ethereum-optimism/optimism/issues/10592) for an example of what this tracking issue should look like.
```

Comment:
Nit: that PR's title has a version number in it. It may actually be heard to predict the final version number once semver is driven by code changes rather than gov cycles. 
Perhaps we should enumerate changelog PRs with their expected governance cycle number, or just "upgrade N".

Reply:
Great point. The expected governance cycle can slip, and upgrade order can change, so something constant is ideal. Perhaps we just give it a descriptive name only? For example, "Contracts: Small improvements and Tech Debt Release Tracker"

Reply:
Clarified this to use upgrade names in 0f9f896